napari/layers/labels/labels.py

```diff
@@ -633,19 +649,33 @@ def _raw_to_displayed(self, raw):
             not self.show_selected_label
             and self._color_mode == LabelColorMode.AUTO
         ):
-            image = np.where(
-                raw > 0, low_discrepancy_image(raw, self._seed), 0
-            )
+            try:
+                image = self._all_vals[raw]
+            except IndexError:
+                max_val = np.max(raw)
+                self._all_vals = low_discrepancy_image(
+                    np.arange(max_val + 1), self._seed
+                )
+                self._all_vals[0] = 0
+                image = self._all_vals[raw]
         elif (
             self.show_selected_label
             and self._color_mode == LabelColorMode.AUTO
         ):
-            selected = self._selected_label
-            image = np.where(
-                raw == selected,
-                low_discrepancy_image(selected, self._seed),
-                0,
+            selected_color = low_discrepancy_image(
+                self._selected_label, self._seed
             )
+            colors = np.zeros(len(self._all_vals))
+            try:
+                colors[self.selected_label] = selected_color
+            except IndexError:
+                max_val = np.max(raw)
+                self._all_vals = low_discrepancy_image(
+                    np.arange(max_val + 1), self._seed
+                )
+                self._all_vals[0] = 0
+                colors = np.zeros(len(self._all_vals))
+            image = colors[raw]
```

Comment:
This seems like it's missing a `colors[self.selected_label] = selected_color`?

And anyway, I think in the *selected* case we don't want to use try/except: we don't need to run np.max, so we can always check `if self.selected_label > len(self._all_vals):`

... right?

Reply:
> This seems like it's missing a colors[self.selected_label] = selected_color?

Yes.

> And anyway, I think in the selected case we don't want to use try/except: we don't need to run np.max, so we can always check if self.selected_label > len(self._all_vals):

I'm actually not strictly sure about this. 

I think we DO want to try/except because it's possible for the user to cycle using the spinbox to a previously unused label and start painting with it. So we have no guarantee that `self.selected_label` has an associated color in `all_vals`.

We could get away with not recomputing `all_vals` and just doing `np.zeros(self.selected_label)`, and let the auto mode try/except redo `all_vals`, but I don't know if that's a better idea. 

What I think this actually should be is this:

```python
            try:
                colors[self.selected_label] = selected_color
            except IndexError:
                self._all_vals = low_discrepancy_image(
                    np.arange(self.selected_label + 1), self._seed
                )
                self._all_vals[0] = 0
                colors = np.zeros(len(self._all_vals))
                colors[self.selected_label] = selected_color
```
If `selected_label` is a new unseen label, then it's the new max, so we wouldn't need to recompute the max, just the colors.